File: algorithm/convert-sorted-array-to-binary-search-tree/convert-sorted-array-to-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:

    TreeNode* buildTree(vector<int>& nums, int l, int r) {
        if (l > r)
            return NULL;
        
        int m = (l + r) / 2;
        TreeNode *root = new TreeNode(nums[m]);
        
        if (l == r)
            return root;
        
        root->left = buildTree(nums, l, m - 1);
        root->right = buildTree(nums, m + 1, r);
        return root;
    }

    TreeNode* sortedArrayToBST(vector<int>& nums) {
        return buildTree(nums, 0, nums.size() - 1);
    }
};

File: algorithm/convert-sorted-array-to-binary-search-tree/convert-sorted-array-to-binary-search-tree.go
package convert_sorted_array_to_binary_search_tree

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func sortedArrayToBST(nums []int) *TreeNode {
	if len(nums) == 0 {
		return nil
	} else if len(nums) == 1 {
		return &TreeNode{Val: nums[0]}
	}

	// size >= 2, e.g.
	// - [3, 4]
	// - [3, 4, 5]

	mid := len(nums) / 2
	// - [3, 4] -> 1
	// - [3, 4, 5] -> 1

	left := nums[:mid]
	// [3]
	// [3]

	right := nums[mid+1:]
	// []
	// [5]

	return &TreeNode{
		Val:   nums[mid],
		Left:  sortedArrayToBST(left),
		Right: sortedArrayToBST(right),
	}
}
